ddprofiler/app/src/main/java/ddprofiler/preanalysis/PreAnalyzer.java

```diff
@@ -130,26 +150,100 @@ public Map<Attribute, Values> readRows(int num) {
+            if (!attribute.getColumnType().equals(AttributeType.UNKNOWN))
                 continue;
-            AttributeType aType;
+
+            AttributeType attributeType;
             if (pc.getBoolean(ProfilerConfig.EXPERIMENTAL)) {
                 // In experimental mode - force all data to be strings
-                aType = AttributeType.STRING;
+                attributeType = AttributeType.STRING;
             } else {
-                aType = typeOfValue(e.getValue());
+                attributeType = typeOfValue(entry.getValue());
             }
-            if (aType == null) {
+            if (attributeType == null) {
                 continue; // Means that data was dirty/anomaly, so skip value
             }
-            a.setColumnType(aType);
-            // a.setColumnType(AttributeType.STRING);
+            attribute.setColumnType(attributeType);
         }
     }
 
+    private void estimateSemanticTypes(Map<Attribute, List<String>> data) {
+        // estimate data type for each attribute
+        for (Entry<Attribute, List<String>> entry : data.entrySet()) {
+            Attribute attribute = entry.getKey();
+            if (attribute.getSemanticType() != null && !attribute.getSemanticType().isEmpty()) {
+                continue;
+            }
+            Map<String, String> semanticType = semanticTypeOfValue(entry.getValue());
+            attribute.setSemanticType(semanticType);
+        }
+    }
+
+    private Map<String, String> semanticTypeOfValue(List<String> values) {
+        Map<String, String> semanticTypes = new HashMap<>();
+        Map<String, Integer> temporalMatchCounts = new HashMap<>();
+        Map<String, Integer> spatialMatchCounts = new HashMap<>();
+
+        for (String value : values) {
+            String granularity = checkTemporalGranularity(value);
+            if (granularity != null) {
+                temporalMatchCounts.put(granularity, temporalMatchCounts.getOrDefault(granularity, 0) + 1);
+                continue;
+            }
+
+            granularity = checkSpatialGranularity(value);
+            if (granularity != null) {
+                spatialMatchCounts.put(granularity, spatialMatchCounts.getOrDefault(granularity, 0) + 1);
+            }
+        }
+
+        if (temporalMatchCounts.isEmpty() && spatialMatchCounts.isEmpty()) {
+            return semanticTypes;
+        }
+
+        // The dummy entry is created to avoid having to check which map is empty
+        Entry<String, Integer> maxTemporalGranularityEntry = temporalMatchCounts.entrySet().stream()
+                .max(Entry.comparingByValue()).orElse(new AbstractMap.SimpleEntry<>("dummy", 0));
+        Entry<String, Integer> maxSpatialGranularityEntry = spatialMatchCounts.entrySet().stream()
+                .max(Entry.comparingByValue()).orElse(new AbstractMap.SimpleEntry<>("dummy", 0));
+
+        if (maxTemporalGranularityEntry.getValue() > maxSpatialGranularityEntry.getValue()) {
```

Comment:
don't need this after separating spatial and temporal

Reply:
Changed at 6010b96.